### PR TITLE
Fixes to mysql modules and state modules

### LINKED ISSUE #60031: Bug #53199 is present in Ubuntu 20.04 MySQL8
**Description**
See #53199. MySQL returns list of privileges. Salt compares to "ALL PRIVILEGES". THis obviously always fails.

**Setup**
Minion:
Stock Ubuntu 20.04

mysql-client-8.0 8.0.23-0ubuntu0.20.04.1
mysql-client-core-8.0 8.0.23-0ubuntu0.20.04.1
mysql-common 5.8+1.0.5ubuntu2
mysql-server 8.0.23-0ubuntu0.20.04.1
mysql-server-8.0 8.0.23-0ubuntu0.20.04.1
mysql-server-core-8.0 8.0.23-0ubuntu0.20.04.1

salt-common 3003+ds-1
salt-minion 3003+ds-1

**Steps to Reproduce the behavior**
```
mailbackup-grants:
   mysql_grants.present:
    - grant: ALL PRIVILEGES
    - database: mailbackup.*
    - user: mailbackup
    - host: localhost
```
**Expected behavior**
Execute state successfully.

**Screenshots**

```
2021-04-14 17:11:50,419 [salt.loaded.int.module.mysql:2444][DEBUG   ][19274] grants mismatch '{'user': '`mailbackup`', 'database': '*.*', 'host': '`localhost`', 'grant': ['USAGE']}'<>'{'user': 'mailbackup', 'host': 'localhost', 'grant': ['ALTER', 'ALTER ROUTINE', 'BACKUP_ADMIN', 'BINLOG_ADMIN', 'CONNECTION_ADMIN', 'CREATE', 'CREATE ROLE', 'CREATE ROUTINE', 'CREATE TABLESPACE', 'CREATE TEMPORARY TABLES', 'CREATE USER', 'CREATE VIEW', 'DELETE', 'DROP', 'DROP ROLE', 'ENCRYPTION_KEY_ADMIN', 'EVENT', 'EXECUTE', 'FILE', 'GROUP_REPLICATION_ADMIN', 'INDEX', 'INSERT', 'LOCK TABLES', 'PERSIST_RO_VARIABLES_ADMIN', 'PROCESS', 'REFERENCES', 'RELOAD', 'REPLICATION CLIENT', 'REPLICATION SLAVE', 'REPLICATION_SLAVE_ADMIN', 'RESOURCE_GROUP_ADMIN', 'RESOURCE_GROUP_USER', 'ROLE_ADMIN', 'SELECT', 'SET_USER_ID', 'SHOW DATABASES', 'SHOW VIEW', 'SHUTDOWN', 'SUPER', 'SYSTEM_VARIABLES_ADMIN', 'TRIGGER', 'UPDATE', 'XA_RECOVER_ADMIN'], 'database': '`mailbackup`.*'}'
```

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
Salt Version:
           Salt: 3000.9

Dependency Versions:
           cffi: 1.9.1
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
         Jinja2: 2.9.4
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: 2.17
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Aug 22 2020, 10:03:02)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.1

System Versions:
           dist: debian 9.13
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-14-amd64
         system: Linux
        version: debian 9.13
```
</details>

Minion is 3003

**Additional context**

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Update alter_db to return True or False depending on the success of failure of the alter.  Update grant_exists to only use the full list of available privileges when the grant is on the global level, eg. datbase is "*.*".  Adding some pytest integration tests, testing across all MySQL variants.

### What issues does this PR fix or reference?
Fixes: #60031 #60264

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by s0undt3ch** [comment]
@garethgreenaway I added two commits to your PR. One to install `pymysql` on linux so test can run, and the other to use pytest-salt-factories container support.

**Review by s0undt3ch — DISMISSED** [review]

File: salt/modules/mysql.py [hunk 1]
```diff
@@ -1110,7 +1110,15 @@ def alter_db(name, character_set=None, collate=None, **connection_args):
...
     )
     args = {}
-    _execute(cur, qry, args)
+    try:
+        if _execute(cur, qry, args):
+            log.info("DB '%s' altered", name)
+            return True
+    except MySQLdb.OperationalError as exc:
+        err = "MySQL Error {}: {}".format(*exc.args)
+        __context__["mysql.error"] = err
```
Comment (s0undt3ch) [inline salt/modules/mysql.py:1119]: Why are we keeping the error in `__context__`? Are we using it somewhere?

File: salt/modules/mysql.py [hunk 2]
```diff
@@ -1134,7 +1142,7 @@ def db_get(name, **connection_args):
         "INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME=%(dbname)s;"
     )
     args = {"dbname": name}
-    _execute(cur, qry, args)
+    ret = _execute(cur, qry, args)
```
Comment (s0undt3ch) [inline salt/modules/mysql.py:1145]: Why assign to `ret` if we're not using it?

File: tests/pytests/unit/modules/test_mysql.py [hunk 3]
```diff
@@ -0,0 +1,922 @@
...
+    with patch.object(mysql, "version", return_value="8.0.10"):
+        mock = MagicMock(return_value=mock_grants)
+        with patch.object(
+            mysql, "user_grants", return_value=mock_grants
+        ) as mock_user_grants:
+            ret = mysql.grant_exists("all privileges", "*.*", "testuser", "%")
+            assert ret
+
+
+@skipIf(True, "TODO: Mock up user_grants()")
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_mysql.py:636]:
```suggestion
@pytest.mark.skipif(True, reason="TODO: Mock up user_grants()")
```

File: tests/pytests/unit/modules/test_mysql.py [hunk 4]
```diff
@@ -0,0 +1,922 @@
...
+def test_grant_add():
+    """
+    Test grant_add function in mysql exec module
+    """
+    _test_call(
+        mysql.grant_add, "", "SELECT,INSERT,UPDATE", "database.*", "frank", "localhost",
+    )
+
+
+@skipIf(True, "TODO: Mock up user_grants()")
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_mysql.py:646]:
```suggestion
@pytest.mark.skipif(True, reason="TODO: Mock up user_grants()")
```

File: tests/pytests/unit/modules/test_mysql.py [hunk 5]
```diff
@@ -0,0 +1,922 @@
...
+    """
+    connect_mock = MagicMock(return_value=None)
+    with patch.object(mysql, "_connect", connect_mock):
+        with patch.dict(mysql.__salt__, {"config.option": MagicMock()}):
+            rslt = mysql.get_slave_status()
+            connect_mock.assert_has_calls([call()])
+            assert rslt == []
+
+
+@skipIf(True, "MySQL module claims this function is not ready for production")
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_mysql.py:694]:
```suggestion
@pytest.mark.skipif(True, reason="MySQL module claims this function is not ready for production")
```

File: tests/pytests/unit/modules/test_mysql.py [hunk 6]
```diff
@@ -0,0 +1,922 @@
...
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""
+
+
+import logging
+
+import pytest
+import salt.modules.mysql as mysql
+from tests.support.mock import MagicMock, call, mock_open, patch
+from tests.support.unit import skipIf
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_mysql.py:15]:
```suggestion
```

File: tests/pytests/unit/modules/test_mysql.py [hunk 7]
```diff
@@ -0,0 +1,922 @@
...
+try:
+    import MySQLdb  # pylint: disable=W0611
+except ImportError:
+    NO_MYSQL = True
+
+try:
+    # MySQLdb import failed, try to import PyMySQL
+    import pymysql
+except ImportError:
+    NO_PyMYSQL = True
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_mysql.py:29]:
```suggestion
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: salt/modules/mysql.py (hunk 1 above)
Comment (garethgreenaway) [reply salt/modules/mysql.py:1119]: That error in `__context__` is used in the state modules.

Comment (s0undt3ch) [reply salt/modules/mysql.py:1119]: :+1:

**Review by s0undt3ch — DISMISSED** [review]

File: requirements/static/ci/linux.in [hunk 8]
```diff
@@ -2,3 +2,6 @@
 pyiface
 pygit2<1.1.0; python_version <= '3.8'
 pygit2>=1.4.0; python_version > '3.8'
+pymysql==0.9.3; python_version <= '3.5'
+pymysql>=1.0.2; python_version > '3.5'
+
```
Comment (Ch3LL) [inline requirements/static/ci/linux.in:7]: remove line

**Review by Ch3LL — CHANGES_REQUESTED** [review]
just one small thing

**Comment by Ch3LL** [comment]
Some failing tests on centos7

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Review by Ch3LL — DISMISSED** [review]

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**